practice_app/api/views.py

```diff
@@ -290,3 +293,119 @@ def semantic_scholar(request):
+    user_id = request.GET.get('user_id')
+
+    Headers = {"Accept": "application/json"}
+
+    if user_id == None or user_id == '':
+        return JsonResponse({"status":"ORCID ID should be provided as user_id"}, status = 404)
+    
+    # third party api call 
+    # returns a json file that contains all public information related with given ORCID ID
+    api_request = requests.get("https://orcid.org/"+user_id, headers=Headers)
+
+    if api_request.status_code != 200:
+        return JsonResponse({"status":"Invalid ORCID ID"}, status = 404)
+    else:
+        try:
+            api_request = api_request.json()
+        except:
+            return JsonResponse({"status":"Invalid ORCID ID"}, status = 404)
+        
+        response = {}
+        response["user_id"] = user_id
+        response["name"] = api_request["person"]["name"]["given-names"]["value"]
+
+        if api_request["person"]["name"]["family-name"] != None:
+            response["surname"] = api_request["person"]["name"]["family-name"]["value"]
+        else:
+            response["surname"] = " "
+        return JsonResponse(response)
+
+# POST api/log_in/
+# implements log in functionality
+# username and password should be provided in Headers
+@csrf_exempt
+def log_in(request):
+    
+    username = request.headers["username"]
+    password = request.headers["password"]
+
+    if username == None or username == '':
+        return JsonResponse({"status":"ORCID ID should be provided."}, status = 404)
+    
+    if password == None or password == '':
+        return JsonResponse({"status":"Password should be provided."}, status = 404)
+    
+    # Check user database to authenticate the user with given credentials. Return a user if valid username and password is given.
+    user = authenticate(request, username=username, password=password)
+
+    # if the user is authenticated log in by built-in login function
+    if user is not None:
+        login(request, user)
+        return JsonResponse({"status":"User logged in."}, status = 200)
+
+    else:
+        return JsonResponse({"status":"Authentication failed"}, status = 404)
+
+# GET api/log_out/
+# get user to log out by built-in logout function
+def log_out(request):
+    logout(request)
+    return JsonResponse({"status":"User logged out."}, status = 200)
+
+# POST api/user_registration
+# username and password should be in headers
+# username should be unique
+# name and surname should be provided in data
+# username, name, password must be provided, surname is optional
+
+@csrf_exempt
+def user_registration(request):
+
+    user_id = request.headers["username"]
+    password = request.headers['password']
+
+    # create a GET request to call orcid_api()
+    orcid_api_request = HttpRequest()
+    orcid_api_request.method = 'GET'
+    orcid_api_request.user = request.user
+    orcid_api_request.META = request.META
+    orcid_api_request.GET.update({"user_id":"0009-0005-5924-1831"})
```

Comment:
this part is hardcoded.

Reply:
I forgot to change after testing the code. Now, it is fixed.